network-mux/demo/cardano-ping.hs

```diff
@@ -46,56 +47,55 @@ keepaliveNum = MiniProtocolNum 8
 
 data Flag = HostF String | PortF String | MagicF String | QuietF | JsonF deriving Show
 
-options :: [OptDescr Flag]
-options = [
-    Option "h" ["h"]     (ReqArg HostF "host")  "hostname/ip, e.g relay.iohk.example",
+optionDescriptions :: [OptDescr Flag]
+optionDescriptions = [
+    Option "h" ["host"]  (ReqArg HostF "host")  "hostname/ip, e.g relay.iohk.example",
     Option "m" ["magic"] (ReqArg MagicF "magic") ("magic, defaults to " ++ show mainnetMagic),
     Option "j" ["json"]  (NoArg JsonF ) "json output flag",
     Option "p" ["port"]  (ReqArg PortF "port") "portnumber, e.g 1234",
     Option "q" ["quiet"] (NoArg QuietF ) "quiet flag, csv/json only output."
   ]
 
-getHost :: [Flag] -> String
-getHost [] = error "Specify host/ip with '-h <hostname>'"
-getHost (HostF h:_) = h
-getHost (_:xs) = getHost xs
 
-getPort :: [Flag] -> String
-getPort [] = error "Specify port with '-p <port number>'"
-getPort (PortF p:_) = p
-getPort (_:xs) = getPort xs
-
-getQuiet :: [Flag] -> Bool
-getQuiet [] = False
-getQuiet (QuietF:_) = True
-getQuiet (_:xs) = getQuiet xs
-
-getMagic :: [Flag] -> Word32
-getMagic [] = mainnetMagic
-getMagic (MagicF m:_) = Prelude.read m
-getMagic (_:xs) = getMagic xs
+data Options = Options {
+      host  :: Maybe String
+    , port  :: String
+    , magic :: Word32
+    , json :: Bool
+    , quiet :: Bool
+    } deriving Show
+
+defaultOpts :: Options
+defaultOpts = Options {
+      host  = Nothing
+    , port  = "3001"
+    , json  = False
+    , quiet = False
+    , magic = mainnetMagic
+    }
 
-getJson :: [Flag] -> Bool
-getJson [] = False
-getJson (JsonF:_) = True
-getJson (_:xs) = getJson xs
+buildOptions ::  Flag -> Options -> Options
+buildOptions (HostF host) opt = opt { host = Just host }
+buildOptions (PortF port) opt = opt { port = port }
+buildOptions (MagicF m)   opt = opt { magic = Prelude.read m }
+buildOptions JsonF        opt = opt { json = True }
+buildOptions QuietF       opt = opt { quiet = True }
 
 main :: IO ()
 main = do
     args <- getArgs
-    let (actions, _, _ ) = getOpt RequireOrder options args
-        host = getHost actions
-        port = getPort actions
-        quiet = getQuiet actions
-        magic = getMagic actions
-        jsonF = getJson actions
+    let (flags, _, _ ) = getOpt RequireOrder optionDescriptions args
+        options = foldr buildOptions defaultOpts flags
         hints = Socket.defaultHints { Socket.addrSocketType = Socket.Stream }
-    addresses <- Socket.getAddrInfo (Just hints) (Just host) (Just port)
 
-    unless jsonF $
+    when (isNothing $ host options) $ error "Specify host/ip with '-h <hostname>'"
```

Comment:
It would be nicer to `putStrlLn` and then throw `ExitFailure 1 :: ExitCode` (or whatever code that fits better).

Reply:
:+1: